src/sql/src/plan/statement/ddl.rs

```diff
@@ -856,26 +856,44 @@ pub fn plan_create_source(
     };
 
     let (available_subsources, requested_subsources) = match (available_subsources, subsources) {
-        (Some(available_subsources), Some(CreateSourceSubsources::Subset(subsources))) => {
+        (Some(available_subsources), Some(CreateReferencedSubsources::Subset(subsources))) => {
             let mut requested_subsources = vec![];
             for subsource in subsources {
-                let (name, target) = match subsource {
-                    CreateSourceSubsource::Resolved(name, target) => (name, target),
-                    CreateSourceSubsource::Aliased(_, _) | CreateSourceSubsource::Bare(_) => {
-                        sql_bail!(
-                            "[internal error] subsources should be resolved during purification"
-                        )
+                let name = subsource.reference.clone();
+
+                let target = match &subsource.subsource {
+                    Some(subsource) => match subsource {
+                        DeferredObjectName::Named(target) => target.clone(),
+                        DeferredObjectName::Deferred(name) => {
+                            // TODO: remove this after the next release, we need
+                            // it only so we can load the catalog to do the
+                            // proper rewrite.
+                            let partial_subsource_name =
+                                normalize::unresolved_object_name(name.clone())?;
+                            let item = scx.catalog.resolve_item(&partial_subsource_name).unwrap();
```

Comment:
Hm, why does the Deferred name looked up in the catalog? I thought that it doesn't exist yet if we're in the planner and we're processing a `Deferred` object.

Some more things are unclear to me, so I'll try to ask with an example.

So in currently deployed catalogs we have sources that use the `name INTO "object"` syntax to hint to the parser that it should parse a raw object name instead of an unqualified one. However this PR parses both `AS` and `INTO` style statements into the `Deferred` variant. And the planner always receives that, regardless of whether or not the target item exists in the catalog. 

What I was expecting was that the parser would have a flag that tells it whether or not the discovered deferred objects should be parsed as `Named` or `Deferred` and that flag to be set by the adapter depending on whether or not it is parsing SQL from user input or from the catalog. This way when a `Statement<Raw>` goes through name resolution we will either ignore `Deferred` objects or try resolve them through the catalog and then the planner will know if it needs to plan just the `CreateSource` item or if it should create subsources too.

Reply:
We are doing this only so we can actually open the catalog to perform the
migration--that's what the `TODO` is above here. This should rightly be an
internal error, but I've lazily made it panic if the object isn't yet created
because we want to be sure this behavior is dangerous.

Regarding the example, there are two code paths we need to satisfy:

**Newly created sources with subsources**

Those will be purified and would only encounter this code path (the match of
`Deferred`) through a programming error. That will panic because the object will
not have been created yet.

**Previously created sources with subsources**

On opening the on-disk catalog to apply migrations, those will be parsed but
_not_ purified before planning. During planning those will encounter this code
path--we have to make sure they're resolvable, obviously.

Your suggestion for adding a flag that lets us toggle whether or not these
should be planned is what my intuition was, as well--however, that toggle needs
to be placed in the AST element itself (`DeferredObjectName::Deferred`), _and_
needs to be parseable (i.e. we cannot just sneak it in as metadata; it needs to
be present the `create_sql` statement on the `CatalogItem::Source`). This is
because we load the catalog for the complex migrations using
`adapter::catalog::parse_item`, which doesn't currently give us a chance to
"inject" this toggle. Adding some kind of post-parsing+pre-planning migration
logic seemed unnecessarily complex.

So to add a toggle, we would need to add some kind of AST element, e.g.
`RESOLVABLE` as a prefix/suffix to the object name, which includes baking the
migration logic into the parser, as well.

I figured that simply letting this code resolve the object, because this only
happens during the catalog opening for the migration, was a cleaner solution
whose previous behavior can return in the next release.

If you'd rather have the AST express this, I'm glad to make the change but was just trying to cut down on the noise/shifting structure here.